bn_s_mp_to_decimal_fast.c

```diff
@@ -0,0 +1,234 @@
+      goto LBL_ERR;
+   }
+
+   if ((err = mp_init_copy(&number, a)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+   if (mp_isneg(&number)) {
+      if ((err = mp_neg(&number, &number)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      result[0] = '-';
+      *result_addr += 1;
+      maxlen -= 1;
+   }
+   mp_set_u32(&n, 1000);
+
+   if ((err = mp_init_copy(&nL[0], &n)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   if ((err = mp_init_set(&shift, (mp_digit)20)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   if ((err = mp_init_copy(&shiftL[0], &shift)) != MP_OKAY) {
+      goto LBL_ERR;
+   }
+
+   /* (8 * 2**$shift) / $n rounded up */
+   mp_set_u32(&M, 8389);
+
+   /* $M / 8, rounded up */
+   mp_set_u32(&mL[0], 1049);
+
+   while (1) {
+      if ((err = mp_sqr(&n, &n)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      if (mp_cmp(&n, &number) == MP_GT) {
+         break;
+      }
+
+      if ((err = mp_mul_2(&shift, &shift)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+
+      /* The following is a Newton-Raphson step, to restore the invariant
+       * that $M is (8 * 2**$shift) / $n, rounded up. */
+      {
+         if ((err = mp_sqr(&M, &M2)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_sqr(&M2, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+
+         if ((err = mp_mul(&M4, &n, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_div_2d(&M4, mp_get_i32(&shift) + 6, &M4, NULL)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_mul_2(&M2, &M2)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_sub(&M4, &M2, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_add_d(&M4, 1, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_div_2d(&M4, 3, &M4, NULL)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_sub_d(&M4, 1, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_neg(&M4, &M)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+      }
+
+      if ((err = mp_init_copy(&nL[precalc_array_index], &n)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      if ((err = mp_init_copy(&shiftL[precalc_array_index], &shift)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+
+      /* Divide by 8, round up */
+      {
+         if ((err = mp_add_d(&M4, 1, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_div_2d(&M4, 3, &M4, NULL)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_sub_d(&M4, 1, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+         if ((err = mp_neg(&M4, &M4)) != MP_OKAY) {
+            goto LBL_ERR;
+         }
+      }
+      if ((err = mp_init_copy(&mL[precalc_array_index], &M4)) != MP_OKAY) {
+         goto LBL_ERR;
+      }
+      precalc_array_index++;
```

Comment:
I don't think that `precalc_array_index` can get larger than 20  but I would prefer a little check here, just in case.

Reply:
why not simply do a `for-loop` instead of the `while(1)` and check the `precalc_array_index` directly after the loop?

as it is now it could still be possible that you're overwriting `nl` and `shiftL`